### PR TITLE
Add additional template test case

### PR DESCRIPTION
**Pull request by llvm-beanz**
Test case authored by @tex3d!

This test case covers a whole bunch of edge conditions around using
templates with builtin types. Huge thank you to @tex3d for writing this
up and giving me a great basis to write fixes based on.

This test case covers a bunch of super common use cases for templates
interacting with builtin types.

### CODE REVIEW

File: tools/clang/test/HLSLFileCheck/hlsl/template/DependentWithBuiltinTemplate.hlsl [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+// BUFFER: %CB0 = type { %"struct.Wrapper<Foo>" }
+// BUFFER: %"struct.Wrapper<Foo>" = type { %struct.Foo }
+
+// WRAPPER: %"hostlayout.$Globals" = type { %"hostlayout.struct.WrappedCB<Foo>" }
+// WRAPPER: %"hostlayout.struct.WrappedCB<Foo>" = type {}
+
+// CHECK: define void @main
+
+// BUFFER: !6 = !{i32 0, %CB0* undef, !"CB0", i32 0, i32 0, i32 1, i32 4, null}
+// WRAPPER: !6 = !{i32 0, %"hostlayout.$Globals"* undef, !"$Globals", i32 0, i32 0, i32 1, i32 0, null}
```
Comment (tex3d) [inline tools/clang/test/HLSLFileCheck/hlsl/template/DependentWithBuiltinTemplate.hlsl:21]: Since `WrappedCB` has no data other than the `ConstantBuffer<>` inside, declaring a global of this type should not actually result in any `$Global` cbuffer use.  I think the fact that this is found is an indication of a bug.  As you can see from `%"hostlayout.struct.WrappedCB<Foo>" = type {}`, the buffer type is empty (as it should be), so ultimately, it should not have been included in the cbuffer resource list at all.  Instead, we should have a cbuffer type that indicates the layout of the ConstantBuffer inside the structure, which appears to be missing, if I'm reading this right.  This isn't necessarily something to fix right here, but the test case is checking for something that should not actually be there, and not checking for something that should, if we were to fix the bug.

I will check what the rest of the DXIL/metadata looks like for this case.

**Review by tex3d — APPROVED** [review]
